crates/modelardb_manager/src/remote.rs

```diff
@@ -150,24 +154,108 @@ impl FlightService for FlightServiceHandler {
     async fn do_action(
         &self,
-        _request: Request<Action>,
+        request: Request<Action>,
     ) -> Result<Response<Self::DoActionStream>, Status> {
-        Err(Status::unimplemented("Not implemented."))
+        let action = request.into_inner();
+        info!("Received request to perform action '{}'.", action.r#type);
+
+        if action.r#type == "RegisterNode" {
+            // Extract the cluster node from the action body.
+            let (url, offset_data) = extract_argument(&action.body)?;
+            let (mode, _offset_data) = extract_argument(offset_data)?;
+
+            let server_mode = ServerMode::from_str(mode).map_err(Status::invalid_argument)?;
+            let cluster_node = ClusterNode::new(url.to_string(), server_mode);
+
+            // Use the metadata manager to persist the node to the metadata database.
+            self.context
+                .metadata_manager
+                .save_cluster_node(&cluster_node)
+                .await
+                .map_err(|error| Status::internal(error.to_string()))?;
+
+            // Use the cluster manager to register the node in memory.
+            self.context
+                .cluster_manager
+                .write()
+                .await
+                .register_node(cluster_node)
+                .map_err(|error| Status::internal(error.to_string()))?;
+
+            // TODO: Return the current database schema as well.
+            let connection_info = self.context.remote_data_folder.connection_info().clone();
+
+            Ok(Response::new(Box::pin(stream::once(async {
+                Ok(FlightResult {
+                    body: connection_info.into(),
+                })
+            }))))
+        } else if action.r#type == "RemoveNode" {
+            let (url, _offset_data) = extract_argument(&action.body)?;
+
+            // Remove the node with the given url from the metadata database.
+            self.context
+                .metadata_manager
+                .remove_cluster_node(url)
+                .await
+                .map_err(|error| Status::internal(error.to_string()))?;
+
+            // Remove the node with the given url from the cluster manager.
+            self.context
+                .cluster_manager
+                .write()
+                .await
+                .remove_node(url)
+                .map_err(|error| Status::internal(error.to_string()))?;
+
+            // Flush the node and kill the process running the node.
+            let mut flight_client = FlightServiceClient::connect(format!("grpc://{url}"))
+                .await
+                .map_err(|error| Status::internal(error.to_string()))?;
+
+            let action = Action {
+                r#type: "KillEdge".to_owned(),
+                body: vec![].into(),
+            };
+
+            // Since the process is killed, the error from the request is ignored.
+            let _ = flight_client.do_action(Request::new(action)).await;
```

Comment:
Maybe retry if the error specifies that the node could not be killed to ensure it is truly gone.

Reply:
Added a TODO to do this later when implementing the retry strategy related functionality for the cluster.